### PR TITLE
refactor: added video torch encoder

### PR DESCRIPTION
**Pull request by JamesTang-jinaai**

### CODE REVIEW

File: encoders/video/VideoTorchEncoder/__init__.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    @batching
+    @as_ndarray
+    def encode(self, data: 'np.ndarray', *args, **kwargs) -> 'np.ndarray':
+        if self.channel_axis != self._default_channel_axis:
+            data = np.moveaxis(data, self.channel_axis, self._default_channel_axis)
+        import torch
+        _input = torch.from_numpy(data.astype('float32'))
+        if self.on_gpu:
+            _input = _input.cuda()
+        print(_input.ndim)
```
Comment (nan-wang) [inline encoders/video/VideoTorchEncoder/__init__.py:71]:
```suggestion
```

File: encoders/video/VideoTorchEncoder/__init__.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        _feature = self._get_features(_input).detach()
+        print(_feature.ndim)
```
Comment (nan-wang) [inline encoders/video/VideoTorchEncoder/__init__.py:73]:
```suggestion
```

File: encoders/video/VideoTorchEncoder/__init__.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        if self.on_gpu:
+            _feature = _feature.cpu()
+        _feature = _feature.numpy()
+        print(_feature.ndim)
```
Comment (nan-wang) [inline encoders/video/VideoTorchEncoder/__init__.py:77]:
```suggestion
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

File: encoders/video/VideoTorchEncoder/tests/test_videotorchencoder.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    if 'JINA_TEST_GPU' in os.environ:
+        metas['on_gpu'] = True
+    return VideoTorchEncoder(metas=metas)
+
+
+def test_encoding_results():
+    input_dim = 224
+    output_dim = 512
+    encoder = get_encoder()
+    test_data = np.random.rand(2, 3, 3, input_dim, input_dim)
```
Comment (nan-wang) [inline encoders/video/VideoTorchEncoder/tests/test_videotorchencoder.py:29]:
```suggestion
    input_dim = 224
    output_dim = 512
    num_frames = 10
    test_data = np.random.rand(2, num_frames, 3, input_dim, input_dim)
```

File: encoders/video/VideoTorchEncoder/__init__.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    def post_init(self):
+        super().post_init()
+        import torchvision.models.video as models
+        if self.pool_strategy is not None:
+            self.pool_fn = getattr(np, self.pool_strategy)
+        self.model = getattr(models, self.model_name)(pretrained=True).eval()
+        self.to_device(self.model)
+
+    def _get_features(self, x):
```
Comment (JoanFM) [inline encoders/video/VideoTorchEncoder/__init__.py:47]: I had never seen this, but I think this model wont work with another model than the pretrained. I think we may need to block the model name to the default one and state it clearly in the readme. It is strange that we rely on this pytorch model to have these layers.

**Comment by YueLiu-jina** [comment]
All test pass locally, but first three tests cause a 'kill' when build online, this may relates to a memory issue. The last test can run online.

**Pull request merged**